parser/result.go

```diff
@@ -530,6 +529,36 @@ func (r *result) asMessageDescriptor(node *ast.MessageNode, syntax syntaxType, h
 	return msgd
 }
 
+func (r *result) addReservedNames(names *[]string, node *ast.ReservedNode, syntax syntaxType, handler *reporter.Handler, alreadyReserved map[string]int) {
+	if len(node.Names) > 0 && syntax == syntaxEditions {
+		nameNodeInfo := r.file.NodeInfo(node.Names[0])
+		_ = handler.HandleErrorf(nameNodeInfo.Start(), `must use identifiers, not string literals, to reserved names with editions`)
+	} else if len(node.Identifiers) > 0 && syntax != syntaxEditions {
+		nameNodeInfo := r.file.NodeInfo(node.Identifiers[0])
+		_ = handler.HandleErrorf(nameNodeInfo.Start(), `must use string literals, not identifiers, to reserved names with proto2 and proto3`)
+	}
+	for _, n := range node.Names {
```

Comment:
Should we put this in an if statement for non editions syntax? Feels like we should ignore it entirely in error messages if string literals aren't supported at all.

Reply:
I re-structured this method to do just this. The branch for editions checks that `Names` is empty and processes `Identifiers`. The other branch, does the opposite, checking that `Identifiers` is empty and processes `Names`.